src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalescingAssignmentTests.cs

```diff
@@ -0,0 +1,1408 @@
+    }
+}
+", expectedOutput: @"
+In GetInt
+0
+").VerifyIL("C.Main()", @"
+{
+  // Code size       55 (0x37)
+  .maxstack  3
+  .locals init (C V_0,
+                int? V_1,
+                int? V_2)
+  IL_0000:  newobj     ""C..ctor()""
+  IL_0005:  stloc.0
+  IL_0006:  ldloc.0
+  IL_0007:  ldfld      ""int? C.f1""
+  IL_000c:  stloc.1
+  IL_000d:  ldloca.s   V_1
+  IL_000f:  call       ""bool int?.HasValue.get""
+  IL_0014:  brtrue.s   IL_002b
+  IL_0016:  ldloc.0
+  IL_0017:  call       ""int C.GetInt()""
+  IL_001c:  newobj     ""int?..ctor(int)""
+  IL_0021:  dup
+  IL_0022:  stloc.2
+  IL_0023:  stfld      ""int? C.f1""
+  IL_0028:  ldloc.2
+  IL_0029:  br.s       IL_002c
+  IL_002b:  ldloc.1
+  IL_002c:  box        ""int?""
+  IL_0031:  call       ""void System.Console.WriteLine(object)""
+  IL_0036:  ret
+}
+");
+        }
+
+        [Fact]
+        public void IndexerLvalue()
+        {
+
+            CompileAndVerify(@"
+using System;
+public class C
+{
+    int? f1 = null;
+
+    public static void Main()
+    {
+        int?[] a = new int?[1];
+        Console.WriteLine(a[GetInt()] ??= 1);
+        Console.WriteLine(a[GetInt()] ??= 2);
+    }
+    static int GetInt()
+    {
+        Console.WriteLine(""In GetInt"");
+        return 0;
+    }
+}
+", expectedOutput: @"
+In GetInt
+1
+In GetInt
+1
+").VerifyIL("C.Main()", @"
+{
+  // Code size      116 (0x74)
+  .maxstack  4
+  .locals init (int?& V_0,
+                int?& V_1,
+                int? V_2)
+  IL_0000:  ldc.i4.1
+  IL_0001:  newarr     ""int?""
+  IL_0006:  dup
+  IL_0007:  call       ""int C.GetInt()""
+  IL_000c:  ldelema    ""int?""
+  IL_0011:  stloc.0
+  IL_0012:  ldloc.0
+  IL_0013:  stloc.1
+  IL_0014:  ldloc.1
+  IL_0015:  call       ""bool int?.HasValue.get""
+  IL_001a:  brtrue.s   IL_002d
+  IL_001c:  ldloc.0
+  IL_001d:  ldc.i4.1
+  IL_001e:  newobj     ""int?..ctor(int)""
+  IL_0023:  dup
+  IL_0024:  stloc.2
+  IL_0025:  stobj      ""int?""
+  IL_002a:  ldloc.2
+  IL_002b:  br.s       IL_0033
+  IL_002d:  ldloc.1
+  IL_002e:  ldobj      ""int?""
+  IL_0033:  box        ""int?""
+  IL_0038:  call       ""void System.Console.WriteLine(object)""
+  IL_003d:  call       ""int C.GetInt()""
+  IL_0042:  ldelema    ""int?""
+  IL_0047:  stloc.1
+  IL_0048:  ldloc.1
+  IL_0049:  stloc.0
+  IL_004a:  ldloc.0
+  IL_004b:  call       ""bool int?.HasValue.get""
+  IL_0050:  brtrue.s   IL_0063
+  IL_0052:  ldloc.1
+  IL_0053:  ldc.i4.2
+  IL_0054:  newobj     ""int?..ctor(int)""
+  IL_0059:  dup
+  IL_005a:  stloc.2
+  IL_005b:  stobj      ""int?""
+  IL_0060:  ldloc.2
+  IL_0061:  br.s       IL_0069
+  IL_0063:  ldloc.0
+  IL_0064:  ldobj      ""int?""
+  IL_0069:  box        ""int?""
+  IL_006e:  call       ""void System.Console.WriteLine(object)""
+  IL_0073:  ret
+}
+");
+        }
+
+        [Fact]
+        public void PropertyLvalue()
+        {
+            CompileAndVerify(@"
+using System;
+public class C
+{
+    int? f1 = null;
+    int? P1
```

Comment:
>int? P1 [](start = 4, length = 7)

Another variant on this theme that would useful is `RefReturningMethod() ??= value;`. Also parameters of various kinds `outParam ??= value;`, etc.
Btw, those test ideas can be captured and addressed in a follow-up PR.